### PR TITLE
Fix the fuzz bug

### PR DESCRIPTION
**Pull request by tcharding**
This PR dose a few things:

- Clear shellhcheck warnings due to dynamic file paths (not sure why this passed CI before)
- Use `say_err` to clear another `shellcheck` warning (that I was unable to repro locally)
- Runs `cargo build` and `cargo test` for all crates instead of treating `fuzz` as a special case

Resolve: #10

### CODE REVIEW

**Comment by tcharding** [comment]
We need to be careful before merging this one. Its a good chance to check all our repos that use this tool to see that they either:

- Use a specific commit hash so we can merge
- Have no bugs so we can increase coverage without breaking

**Comment by apoelstra** [comment]
If anything is *not* using a specific commit hash then we need to change them to do so. We can't paralyze this repo by having its `master` branch automatically deployed to a dozen other projects.

**Comment by apoelstra** [comment]
In 3c31ce7794d0b12fa3c2a54edbf43e193ea7ff80:

This fix is fine though I think we should just stop treating the name `fuzz` specially. I don't treat it specially in my local CI setup. I look for honggfuzz deps and run `cargo hfuzz` on all binaries in crates that they appear (and recently, I even choose the correct version of honggfuzz based on the lockfile).

CI here is failing. I think we should fix it.

**Comment by tcharding** [comment]
You were reviewing while I was hacking :)

**Comment by tcharding** [comment]
> If anything is not using a specific commit hash then we need to change them to do so. We can't paralyze this repo by having its master branch automatically deployed to a dozen other projects.

Agreed, I'll check the users to make sure this is the case.

**Comment by tcharding** [comment]
I'm going to come back to this, its more important to check all the users. (And I'm confused why `shellcheck -x foo` is warning, that is a sign to come back with a fresh brain.

**Comment by tcharding** [comment]
Only `miniscript` was missing the use of `rev` (https://github.com/rust-bitcoin/rust-miniscript/pull/720).

**Comment by tcharding** [comment]
> This fix is fine though I think we should just stop treating the name fuzz specially.

I had a go doing this but if we treat it like other crates it requires a `contrib/test_vars.sh` file which seems clunky.

**Comment by tcharding** [comment]
I'm not sure what is going on with the CI job, `shellcheck ci/run_task.sh` runs cleanly for me locally. Can you check on your machine please @apoelstra

**Comment by apoelstra** [comment]
> I had a go doing this but if we treat it like other crates it requires a `contrib/test_vars.sh` file which seems clunky.

Doing this would correctly capture the fact that some of our fuzz harnesses still have `honggfuzz` features while others don't.

But in any case, it seems less clunky to me to have a `contrib/test_vars.sh` than to have a particular workspace which we special-case based on its directory name, and don't even try to run the unit tests on.

**Comment by apoelstra** [comment]
@tcharding when I run shellcheck locally the error appears. It is complaining that the `say_err` function is defined but never called.

We probably want to globally whitelist "unused code" type warnings on this repo.

**Comment by tcharding** [comment]
Cheers, I found a place that we were using `echo` to print an error message so I used `say_err`. I'm still unsure why my `shellcheck` was not showing the warning.

**Comment by tcharding** [comment]
> We probably want to globally whitelist "unused code" type warnings on this repo.

Not done but I'll keep it in mind.

**Review by apoelstra — APPROVED** [review]
ACK f56e456ed282a2a69ca140a0d6734259764f36bb though I still think we should stop special-casing

**Comment by tcharding** [comment]
Can we merge this and I'll work on stopping treating `fuzz` specially, as well as allowing all crates in the `bitcoin` repo to be made part of the workspace.

**Comment by tcharding** [comment]
Total re-write, please not I'm not exactly sure why the last patch works but it does when tested with https://github.com/rust-bitcoin/rust-bitcoin/pull/3201

**Review by apoelstra — APPROVED** [review]
ACK 6646692a37c06388b58c8177d7ee5add1bce313a

File: ci/run_task.sh [hunk 1]
```diff
@@ -146,7 +143,7 @@ do_test() {
     $cargo build
     $cargo test
 
-    if [ -n "${EXAMPLES+x}" ]; then
+    if [ -n "${EXAMPLES}" ]; then
```
Comment (apoelstra) [inline ci/run_task.sh:146]: In 6646692a37c06388b58c8177d7ee5add1bce313a:

What were these `+x`s supposed to mean? AFAIK they just ensure that the `-n` check will always pass and the `-z` checks never will.

Comment (tcharding) [reply ci/run_task.sh:146]: Correct, and also I was annoyed at myself for writing syntax that I had to look  up when I tried to read it. Hence I added the more brain-dead set the vars to `""` in the loop.

**Pull request merged**